File: answers/NodeJS/@CFCIfe/20-love-calculator.js
var http = require("http");
var readlineSync = require("readline-sync");

count = string => {
  return string.split("").reduce((acc, el) => {
    if (acc.hasOwnProperty(el)) acc[el]++;
    else acc[el] = 1;
    return acc;
  }, {});
};

function loveCalc() {
  let firstName = readlineSync.question("First name: "),
    secondName = readlineSync.question("Second name: "),
    name = firstName.toLowerCase() + "loves" + secondName.toLowerCase(),
    nameObj = count(name),
    nameArr = name.split(""),
    objCount = [];

  for (let i = 0; i < nameArr.length; i++) {
    let occurArr = nameObj[nameArr[i]];
    objCount.push(occurArr);
  }
  return parseInt(objCount.join(""), 10);
}

function findPercent() {
  let num = loveCalc();
  let arrNum = num
    .toString()
    .split("")
    .map(item => parseInt(item));

  function sumArray(array) {
    let sum = 0,
      numLatest = [];

    if (array.length % 2 === 0) {
      for (
        let x = 0, y = array.length - 1;
        x <= Math.floor((array.length - 1) / 2) &&
        y >= Math.ceil((array.length - 1) / 2);
        x++, y--
      ) {
        sum = array[x] + array[y];
        numLatest.push(sum);
      }
    } else {
      let middleNum = array[Math.ceil((array.length - 1) / 2)];
      for (
        let x = 0, y = array.length - 1;
        x < Math.floor((array.length - 1) / 2) &&
        y > Math.ceil((array.length - 1) / 2);
        x++, y--
      ) {
        sum = array[x] + array[y];
        numLatest.push(sum);
      }
      numLatest = [numLatest, middleNum];
    }
    numLatest =
      numLatest[0].length > 1
        ? numLatest.flat()
        : [
            ...numLatest
              .join("")
              .split("")
              .map(item => parseInt(item))
          ];
    return numLatest;
  }
  while (arrNum.length > 2) {
    arrNum = sumArray(arrNum);
  }
  finalAnswer = arrNum.join("");
  return `The love percentage is ${finalAnswer}%`;
}

http
  .createServer(function(req, res) {
    res.writeHead(200, { "Content-Type": "text/plain" });
    res.end(findPercent());
  })
  .listen(8080, "127.0.0.1");
console.log("Server running at http://127.0.0.1:8080/");

File: programming/answers/Python/@oseme-techguy/20-love-calculator.py
"""
    Solution to Love Calculator
"""

import math

def reduce_occurrences(occurrences_array):
    size = len(occurrences_array)
    if size <= 2:
        first = occurrences_array[0] if len(occurrences_array) > 0 else ''
        second = occurrences_array[1] if len(occurrences_array) > 1 else ''
        return '{first}{second}'.format(first=first,second=second)

    mid_section = math.floor(size/2)
    new_occurrences_str = ''
    i = 0
    while i < mid_section:
        new_occurrences_str += str(int(occurrences_array[i]) + int(occurrences_array[size - 1 - i]))
        i += 1

    if size % 2 == 1: # size is odd
        new_occurrences_str += str(occurrences_array[i])

    return list(new_occurrences_str)


if __name__ == '__main__':
    while True:
        first_name = input('Enter the first name: ')
        first_name = str(first_name)

        second_name = input('Enter the second name: ')
        second_name = str(second_name)

        joined_names = '{name1}loves{name2}' \
            .format(name1=first_name,name2=second_name) \
            .lower()

        occurrences_array = []

        i = 0
        while i < len(joined_names):
            occurrences_array.append(joined_names.count(joined_names[i]))
            i += 1

        print(occurrences_array)
        result = reduce_occurrences(occurrences_array)
        while type(result) is not str:
            result = reduce_occurrences(result)

        print('\nThe love percentage is {percentage}%\n'.format(percentage=result))

